didman/didman.go

```diff
@@ -0,0 +1,153 @@
+/*
+ * Nuts node
+ * Copyright (C) 2021 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package didman
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"net/url"
+
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"github.com/shengdoushi/base58"
+)
+
+// ModuleName contains the name of this module: Didman
+const ModuleName = "Didman"
+
+// ErrDuplicateService is returned when a DID Document already contains a service for the given type
+var ErrDuplicateService = errors.New("service type already defined")
+
+// ErrServiceInUse is returned when a service is deleted but in use by other services
+var ErrServiceInUse = errors.New("service is referenced by 1 or more services")
+
+// ErrServiceNotFound is returned when the service is not found on a DID
+var ErrServiceNotFound = errors.New("service not found in DID Document")
+
+type didman struct {
+	docResolver types.DocResolver
+	store       types.Store
+	vdr         types.VDR
+}
+
+// NewDidmanInstance creates a new didman instance with services set
+func NewDidmanInstance(docResolver types.DocResolver, store types.Store, vdr types.VDR) Didman {
+	return &didman{
+		docResolver: docResolver,
+		store:       store,
+		vdr:         vdr,
+	}
+}
+
+func (d *didman) Name() string {
+	return ModuleName
+}
+
+func (d *didman) AddEndpoint(id did.DID, serviceType string, u url.URL) error {
+	doc, meta, err := d.docResolver.Resolve(id, nil)
+	if err != nil {
+		return err
+	}
+
+	// check for duplicate service type
+	for _, s := range doc.Service {
+		if s.Type == serviceType {
+			return ErrDuplicateService
+		}
+	}
+
+	// construct service with correct ID
+	service := constructService(id, serviceType, u)
+	doc.Service = append(doc.Service, service)
+
+	return d.vdr.Update(id, meta.Hash, *doc, nil)
+}
+
+func (d *didman) DeleteService(serviceID ssi.URI) error {
+	id, err := did.ParseDID(serviceID.String())
+	if err != nil {
+		return err
+	}
+	id.Fragment = ""
+
+	doc, meta, err := d.docResolver.Resolve(*id, nil)
+	if err != nil {
+		return err
+	}
+
+	// check for existing use
+	if err = d.store.Iterate(func(doc did.Document, metadata types.DocumentMetadata) error {
```

Comment:
will this be performant enough?

Reply:
No, but adding indices on the in-memory implementation would be a waste of time. When the implementation changes to a persistent store, that store hopefully has indices.